extensions/oplogPopulator/modules/Allocator.js

```diff
@@ -0,0 +1,123 @@
+const joi = require('joi');
+const { errors } = require('arsenal');
+const LeastFullConnector = require('../allocationStrategy/LeastFullConnector');
+
+const paramsJoi = joi.object({
+    connectorsManager: joi.object().required(),
+    logger: joi.object().required(),
+}).required();
+
+/**
+ * @class Allocator
+ *
+ * @classdesc Allocator handles listening to buckets by assigning
+ * a connector to them.
+ */
+class Allocator {
+
+    /**
+     * @constructor
+     * @param {Object} params Allocator param
+     * @param {ConnectorsManager} params.connectorsManager connectorsManager
+     * @param {Logger} params.logger logger object
+     */
+    constructor(params) {
+        joi.attempt(params, paramsJoi);
+        this._connectorsManager = params.connectorsManager;
+        this._logger = params.logger;
+        this._allocationStrategy = new LeastFullConnector({
+            logger: params.logger,
+        });
+        // Stores connector assigned for each bucket
+        this._bucketsToConnectors = new Map();
+        this._initConnectorToBucketMap();
+    }
+
+    /**
+     * Initializes connectorsManager map
+     * @returns {undefined}
+     */
+    _initConnectorToBucketMap() {
+        const connectors = this._connectorsManager.connectors;
+        connectors.forEach(connector => {
+            connector.buckets
+                .forEach(bucket => this._bucketsToConnectors.set(bucket, connector));
+        });
+    }
+
+    /**
+     * Is bucket already added to a connector
+     * @param {string} bucket bucket name
+     * @returns {boolean} true if a connector is
+     * listening to the bucket
+     */
+    has(bucket) {
+        return this._bucketsToConnectors.has(bucket);
+    }
+
+    /**
+     * Starts listening to bucket by
+     * adding and assigning a connector to it
+     * @param {string} bucket bucket name
+     * @param {boolean} isRetry is function call a retry
+     * @returns {Promise|undefined} undefined
+     * @throws {InternalError}
+     */
+    async listenToBucket(bucket, isRetry = false) {
+        try {
+            if (!this._bucketsToConnectors.has(bucket)) {
+                const connectors = this._connectorsManager.connectors;
+                const connector = this._allocationStrategy.getConnector(connectors);
+                await connector.addBucket(bucket);
+                this._bucketsToConnectors.set(bucket, connector);
+            }
+        } catch (err) {
+            // We allow one retry as in initial setup
+            // the connector might not be fully ready
+            if (!isRetry) {
```

Comment:
One retry may not always be the best value, we can simply make it more flexible without extra complexity by passing `remainingRetries` instead, decrementing it at each call and stopping when reaching 0 (and then, choosing a reasonable initial value that could be 1 but also something else configurable).

Alternatively, `async.retry` provides support for such pattern and may also be tuned to do things like exponential wait times between retries to cope with transient issues - maybe not useful here but for future reference.